### PR TITLE
[WPE] Support headless command line switch for MiniBrowser with new platform API

### PR DESCRIPTION
**Pull request by lauromoura**
#### a08a4557b5804a6aae9a31e8d4e3a3eb622355f7
<pre>
[WPE] Support headless command line switch for MiniBrowser with new platform API
<a href="https://bugs.webkit.org/show_bug.cgi?id=275249">https://bugs.webkit.org/show_bug.cgi?id=275249</a>

Reviewed by Carlos Garcia Campos.

Create headless display if using `--headless` switch like we used to do
with the old WPE API.

With this, headless webdriver tests get almost on part with the wayland
backend. Some tests time out due to missing WPEKeymap support in the
Headless platform, to be added in a future commit.

* Tools/MiniBrowser/wpe/main.cpp:
(activate): Create a new headless display if requested by the user

Canonical link: <a href="https://commits.webkit.org/280111@main">https://commits.webkit.org/280111@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/702c424bbba6ba40e52e110a8afa80eaa444860f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55732 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35055 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58716 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6163 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57858 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42677 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6361 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44882 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4245 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32958 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48040 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26014 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29745 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5369 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4306 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51717 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60307 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5770 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52308 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48110 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51802 "Found 3 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/http-test-page-list, /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure (failure)") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8229 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30886 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31971 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33052 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31718 "Built successfully") | | |

### CODE REVIEW

**Comment by lauromoura** [comment]
`run-minibrowser` in headless+new api mode seems to be crashing under flatpak, though, regardless of this commit

File: Tools/MiniBrowser/wpe/main.cpp [hunk 1]
```diff
@@ -31,6 +31,10 @@
 #include <memory>
 #include <wpe/webkit.h>
 
+#if ENABLE_WPE_PLATFORM
```
Comment (carlosgcampos) [inline Tools/MiniBrowser/wpe/main.cpp]:
```suggestion
#if ENABLE_WPE_PLATFORM_HEADLESS
```

File: Tools/MiniBrowser/wpe/main.cpp [hunk 2]
```diff
@@ -57,6 +61,7 @@ static gboolean printVersion;
 static GHashTable* openViews;
 #if ENABLE_WPE_PLATFORM
 static gboolean useWPEPlatformAPI;
+static WPEDisplay* wpeDisplay;
```
Comment (carlosgcampos) [inline Tools/MiniBrowser/wpe/main.cpp]: I don't think we need this to be global

File: Tools/MiniBrowser/wpe/main.cpp [hunk 3]
```diff
@@ -418,6 +423,9 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
         "user-content-manager", userContentManager,
         "is-controlled-by-automation", automationMode,
         "website-policies", defaultWebsitePolicies,
+#if ENABLE_WPE_PLATFORM
```
Comment (carlosgcampos) [inline Tools/MiniBrowser/wpe/main.cpp]:
```suggestion
#if ENABLE_WPE_PLATFORM_HEADLESS
```

File: Tools/MiniBrowser/wpe/main.cpp [hunk 4]
```diff
@@ -418,6 +423,9 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
...
+        "display", wpeDisplay,
+#endif
```
Comment (carlosgcampos) [inline Tools/MiniBrowser/wpe/main.cpp:431]: We can create the display before the web view and unref after it, since the view takes a reference of the display.

File: Tools/MiniBrowser/wpe/main.cpp [hunk 5]
```diff
@@ -524,6 +532,9 @@ int main(int argc, char *argv[])
         return 0;
     }
 
+    if (headlessMode && useWPEPlatformAPI && !g_getenv("WPE_DISPLAY"))
+        wpeDisplay = wpe_display_headless_new();
+
```
Comment (carlosgcampos) [inline Tools/MiniBrowser/wpe/main.cpp]: This should be protected by `ENABLE_WPE_PLATFORM_HEADLESS` too. I don't think we want to honor the env var when --headless is passed

File: Tools/MiniBrowser/wpe/main.cpp (hunk 4 above)
Comment (lauromoura) [reply Tools/MiniBrowser/wpe/main.cpp:431]: Right, I'll update it.

File: Tools/MiniBrowser/wpe/main.cpp (hunk 5 above)
Comment (lauromoura) [reply Tools/MiniBrowser/wpe/main.cpp]: I went with the rationale that tweaking env vars would take precedence over command line options, but looks like other tools like `gcc` make cli switches (like `-L`) higher priority indeed.

**Comment by lauromoura** [comment]
Updated the patch, alongside some input method small checks to avoid segfaults.

File: Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp [hunk 6]
```diff
@@ -940,8 +940,10 @@ static void webkitWebViewConstructed(GObject* object)
 #elif PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
     if (priv->display) {
         GRefPtr<WebKitInputMethodContext> imContext = adoptGRef(webkitInputMethodContextImplWPENew(priv->view->wpeView()));
-        webkitInputMethodContextSetWebView(imContext.get(), webView);
-        priv->view->setInputMethodContext(imContext.get());
+        if (imContext) {
+            webkitInputMethodContextSetWebView(imContext.get(), webView);
+            priv->view->setInputMethodContext(imContext.get());
+        }
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: Oh, I didn't realize this was possible when reviewed the IME patch. This should be fixed in a different patch. We should add a None implementation as a fallback for platforms not supporting/implementing IME.

File: Tools/MiniBrowser/wpe/main.cpp [hunk 7]
```diff
@@ -404,6 +408,12 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
         delete static_cast<WPEToolingBackends::ViewBackend*>(data);
     }, backend) : nullptr;
 
+#if ENABLE_WPE_PLATFORM_HEADLESS
+    WPEDisplay* wpeDisplay = nullptr;
+    if (headlessMode && useWPEPlatformAPI)
+        wpeDisplay = wpe_display_headless_new();
```
Comment (carlosgcampos) [inline Tools/MiniBrowser/wpe/main.cpp]:
```suggestion
    WPEDisplay* wpeDisplay = headlessMode && useWPEPlatformAPI ? wpe_display_headless_new() : nullptr;
```

File: Tools/MiniBrowser/wpe/main.cpp [hunk 8]
```diff
@@ -418,9 +428,16 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
...
         "website-policies", defaultWebsitePolicies,
+#if ENABLE_WPE_PLATFORM_HEADLESS
+        "display", wpeDisplay,
+#endif
         nullptr));
     g_object_unref(settings);
     g_object_unref(defaultWebsitePolicies);
+#if ENABLE_WPE_PLATFORM_HEADLESS
+    if (wpeDisplay)
+        g_object_unref(wpeDisplay);
```
Comment (carlosgcampos) [inline Tools/MiniBrowser/wpe/main.cpp]:
```suggestion
    g_clear_object(&wpeDisplay);
```

File: Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp (hunk 6 above)
Comment (lauromoura) [reply Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: Thanks. Already checked https://commits.webkit.org/280009@main and indeed I'm getting no crashes regarding IM.

File: Tools/MiniBrowser/wpe/main.cpp (hunk 7 above)
Comment (lauromoura) [reply Tools/MiniBrowser/wpe/main.cpp]: Added to the current version.

File: Tools/MiniBrowser/wpe/main.cpp (hunk 8 above)
Comment (lauromoura) [reply Tools/MiniBrowser/wpe/main.cpp]: Nice catch. thanks

**Review by carlosgcampos — APPROVED** [review]
